lib/pbench/agent/utils.py

```diff
@@ -14,6 +17,181 @@
+        kill code and a return value.
+
+        A kill code of 0 and return value of 42 is returned as 42.
+        A kill code of 5 and return value of 52 is returned as 542.
+        """
+        return (kill_code * 100) + ret_val
+
+
+class BaseServer:
+    """BaseServer - abstract base class for common code shared between the
+    ToolDataSink and RedisServer classes.
+    """
+
+    def_port = None
+    bad_port_code = None
+    bad_host_code = None
+    name = None
+
+    class Err(BaseReturnCode.Err):
+        """BaseServer.Err - derived from BaseReturnCode.Err, specifically
+        raised by BaseServer and its derived classes.
+        """
+
+        pass
+
+    def __init__(self, spec: str, def_host_name: str):
+        """__init__ - from the given IP/port specification given, determine the
+        IP:port for binding (listening) and the IP:port for connecting.
+
+        The IP/port specification can be given in one of two forms:
+
+          - `<ip>:<port>'
+            * where the same ip address and port are used for binding and
+              connecting
+          - `<bind ip>:<port>;<connect ip>:<port>`
+            * where a semi-colon separates the bind ip/port from the connecting
+              ip/port
+
+        In either case, a missing port (bare colon, optional) indicates the
+        default port should be used. If no IP address is given, the default
+        host name is used.
+
+        No attempt is made to verify that the IP address resolves, or that it
+        is reachable, though we do check they are syntactically valid.
+        """
+        _spec = spec if spec else def_host_name
+        parts = _spec.split(";", 1)
+        pairs = []
+        for part in parts:
+            host_port_parts = part.rsplit(":", 1)
+            if len(host_port_parts) == 1:
+                port = self.def_port
+            else:
+                try:
+                    port = int(host_port_parts[1])
+                except ValueError as exc:
+                    if host_port_parts[1] == "":
+                        port = self.def_port
+                    else:
+                        raise self.Err(
+                            f"Bad port specified for {self.name} in '{spec}'",
+                            self.bad_port_ret_code,
+                        ) from exc
+            host = host_port_parts[0] if host_port_parts[0] else def_host_name
+            if host[0] == "[" and host[-1] == "]":
+                # Brackets are invalid for a host name, but might be used when
+                # specifying a port with an IPv6 address, strip them before we
+                # validate the host name.
+                host = host[1:-1]
+            if validate_hostname(host) != 0:
+                raise self.Err(
+                    f"Bad host specified for {self.name} in '{spec}'",
+                    self.bad_host_ret_code,
+                )
+            pairs.append((host, port))
+
```

Comment:
Is it worth making this a function and adding an all-possible-permutations test for it?

Reply:
Yes, good idea, I should add proper unit tests for this refactored code.

Reply:
Done.